File: FbxSharpTests/AnimCurveKeyTest.cs
using System;
using NUnit.Framework;
using FbxSharp;

namespace FbxSharpTests
{
    [TestFixture]
    public class AnimCurveKeyTest : TestBase
    {
        [Test]
        public void AnimCurveKey_Create_HasDefaultValues()
        {
            // when:
            var key = new FbxAnimCurveKey();

            // then:
            Assert.AreEqual(0L, key.GetTime().Get());
            Assert.AreEqual(0.0, key.GetValue());
            Assert.AreEqual(FbxAnimCurveDef.ETangentMode.eTangentAuto, key.GetTangentMode());
            Assert.AreEqual(FbxAnimCurveDef.EInterpolationType.eInterpolationCubic, key.GetInterpolation());
            Assert.AreEqual(FbxAnimCurveDef.EWeightedMode.eWeightedNone, key.GetTangentWeightMode());
            Assert.AreEqual(FbxAnimCurveDef.EConstantMode.eConstantNext, key.GetConstantMode());
            Assert.AreEqual(FbxAnimCurveDef.EVelocityMode.eVelocityNone, key.GetTangentVelocityMode());
            Assert.AreEqual(FbxAnimCurveDef.ETangentVisibility.eTangentShowNone, key.GetTangentVisibility());
            Assert.True(!key.GetBreak());
        }

        [Test]
        public void AnimCurveKey_SetWeightLeftThenRight_WeightIsAll()
        {
            // given:
            var key = new FbxAnimCurveKey();

            // require:
            Assert.AreEqual(FbxAnimCurveDef.EWeightedMode.eWeightedNone, key.GetTangentWeightMode());

            // when:
            key.SetTangentWeightMode(FbxAnimCurveDef.EWeightedMode.eWeightedNextLeft);

            // then:
            Assert.AreEqual(FbxAnimCurveDef.EWeightedMode.eWeightedNextLeft, key.GetTangentWeightMode());

            // when:
            key.SetTangentWeightMode(FbxAnimCurveDef.EWeightedMode.eWeightedRight, FbxAnimCurveDef.EWeightedMode.eWeightedRight);

            // then:
            Assert.AreEqual(FbxAnimCurveDef.EWeightedMode.eWeightedAll, key.GetTangentWeightMode());
        }

        [Test]
        public void AnimCurveKey_SetTangentWeights_TangentWeightsGetSet()
        {
            // given:
            var key = new FbxAnimCurveKey();
            key.SetTangentWeightMode(FbxAnimCurveDef.EWeightedMode.eWeightedAll);

            // require:
            Assert.AreEqual(FbxAnimCurveDef.EWeightedMode.eWeightedAll, key.GetTangentWeightMode());
            Assert.AreEqual(FbxAnimCurveDef.sDEFAULT_WEIGHT, key.GetDataFloat(FbxAnimCurveDef.EDataIndex.eRightWeight));
            Assert.AreEqual(FbxAnimCurveDef.sDEFAULT_WEIGHT, key.GetDataFloat(FbxAnimCurveDef.EDataIndex.eNextLeftWeight));

            // when:
            key.SetTangentWeightAndAdjustTangent(FbxAnimCurveDef.EDataIndex.eRightWeight, 0.234f);

            // then:
            Assert.AreEqual(0.234f, key.GetDataFloat(FbxAnimCurveDef.EDataIndex.eRightWeight), 0.0001f);
            Assert.AreEqual(FbxAnimCurveDef.sDEFAULT_WEIGHT, key.GetDataFloat(FbxAnimCurveDef.EDataIndex.eNextLeftWeight));

            // when:
            key.SetTangentWeightAndAdjustTangent(FbxAnimCurveDef.EDataIndex.eNextLeftWeight, 0.567f);

            // then:
            Assert.AreEqual(0.234f, key.GetDataFloat(FbxAnimCurveDef.EDataIndex.eRightWeight), 0.0001f);
            Assert.AreEqual(0.567f, key.GetDataFloat(FbxAnimCurveDef.EDataIndex.eNextLeftWeight), 0.0001f);
        }
    }
}

File: FbxCppTests/AnimCurveKeyTest.cpp

#include "Tests.h"

using namespace std;

void AnimCurveKey_Create_HasDefaultValues()
{
    // when:
    FbxAnimCurveKey* key = new FbxAnimCurveKey();

    // then:
    AssertEqual(0LL, key->GetTime().Get());
    AssertEqual(0.0, key->GetValue());
    AssertEqual(FbxAnimCurveDef::eTangentAuto, key->GetTangentMode());
    AssertEqual(FbxAnimCurveDef::eInterpolationCubic, key->GetInterpolation());
    AssertEqual(FbxAnimCurveDef::eWeightedNone, key->GetTangentWeightMode());
    AssertEqual(FbxAnimCurveDef::eConstantNext, key->GetConstantMode());
    AssertEqual(FbxAnimCurveDef::eVelocityNone, key->GetTangentVelocityMode());
    AssertEqual(FbxAnimCurveDef::eTangentShowNone, key->GetTangentVisibility());
    AssertTrue(!key->GetBreak());
}

void AnimCurveKey_SetWeightLeftThenRight_WeightIsAll()
{
    // given:
    FbxManager* manager = FbxManager::Create();
    FbxAnimCurveKey* key = new FbxAnimCurveKey();

    // require:
    AssertEqual(FbxAnimCurveDef::eWeightedNone, key->GetTangentWeightMode());

    // when:
    key->SetTangentWeightMode(FbxAnimCurveDef::eWeightedNextLeft);

    // then:
    AssertEqual(FbxAnimCurveDef::eWeightedNextLeft, key->GetTangentWeightMode());

    // when:
    key->SetTangentWeightMode(FbxAnimCurveDef::eWeightedRight, FbxAnimCurveDef::eWeightedRight);

    // then:
    AssertEqual(FbxAnimCurveDef::eWeightedAll, key->GetTangentWeightMode());
}

void AnimCurveKey_SetTangentWeights_TangentWeightsGetSet()
{
    // given:
    FbxManager* manager = FbxManager::Create();
    FbxAnimCurveKey* key = new FbxAnimCurveKey();
    key->SetTangentWeightMode(FbxAnimCurveDef::eWeightedAll);

    // require:
    AssertEqual(FbxAnimCurveDef::eWeightedAll, key->GetTangentWeightMode());
    AssertEqual(FbxAnimCurveDef::sDEFAULT_WEIGHT, key->GetDataFloat(FbxAnimCurveDef::eRightWeight));
    AssertEqual(FbxAnimCurveDef::sDEFAULT_WEIGHT, key->GetDataFloat(FbxAnimCurveDef::eNextLeftWeight));

    // when:
    key->SetTangentWeightAndAdjustTangent(FbxAnimCurveDef::eRightWeight, 0.234f);

    // then:
    AssertEqual(0.234f, key->GetDataFloat(FbxAnimCurveDef::eRightWeight), 0.0001f);
    AssertEqual(FbxAnimCurveDef::sDEFAULT_WEIGHT, key->GetDataFloat(FbxAnimCurveDef::eNextLeftWeight));

    // when:
    key->SetTangentWeightAndAdjustTangent(FbxAnimCurveDef::eNextLeftWeight, 0.567f);

    // then:
    AssertEqual(0.234f, key->GetDataFloat(FbxAnimCurveDef::eRightWeight), 0.0001f);
    AssertEqual(0.567f, key->GetDataFloat(FbxAnimCurveDef::eNextLeftWeight), 0.0001f);
}

void AnimCurveKeyTest::RegisterTestCases()
{
    AddTestCase(AnimCurveKey_Create_HasDefaultValues);
    AddTestCase(AnimCurveKey_SetWeightLeftThenRight_WeightIsAll);
    AddTestCase(AnimCurveKey_SetTangentWeights_TangentWeightsGetSet);
}

